build: per convention, change default make goal to `build`

benesch opened the pull request:
To adhere to Make convention, change the default Make goal from `build
test lint` to just `build`.

This is another step in addressing #15862.

---

I could have sworn that somewhere we mention that a bare `make` will build, run the tests, and lint, but now that I'm looking for it I can't find it.

benesch committed:
build: per convention, change default make goal to `build`

To adhere to Make convention, change the default Make goal from `build
test lint` to just `build`. Also, remove the `short` target, which ran
`build testshort lintshort`, because it's the short variant of the
behavior we just removed. `testshort` and `lintshort` can continue to be
listed explicilty as targets.

This is another step in addressing #15862.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15909)
<!-- Reviewable:end -->

knz commented on Makefile:
```diff
@@ -128,15 +128,12 @@ endif
 
-.DEFAULT_GOAL := all
-.PHONY: all
-all: build test lint
```
keep it simple:

- don't use DEFAULT_GOAL; `make` by definition uses the first rule defined as default
- just write `all: build` and be done with it (no need for PHONY either, nor the dependency below -- it's implied by all:build)

benesch replied:
We include build/common.mk above, so the first rule has already been defined by the time we get here. We could put an additional target above the include $(REPO_ROOT)/build/common.mk line, but I'm not sure that buys us anything. If you're here, you've already waded through ~100 lines of platform-specific drudgery, so you're probably not afraid of Make.

knz replied:
ok for DEFAULT_GOAL in light of the rules in common.mk

I still think that the PHONY spec is not needed though... oh well.

benesch replied:
Wait, which `PHONY` are you worried about? The one on `build` below is necessary, as is the one in the archive Makefile, no?

knz replied:
I'm not worried, I just think it's not necessary. If you have PHONY on `build` (because there's a directory called `build` like the rule, that's the only point of using PHONY), then any rule that depends on `build` (e.g. `all`) will be phony too. That's my understanding.

benesch replied:
Right—but this change _removes_ the `all` target! That's why I'm confused.

knz replied:
Huh I had missed that. Why remove `all`? Isn't that ... also a common idiom?

knz replied:
(I mean here I had expected that `all` == `build`)

benesch replied:
Good point. Done.

benesch commented:
Just pushed a change to also remove `make short`, which no longer makes sense in a world where `make` doesn't run `test` or `lint`.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15909#-:-KkBMwYtz1551nZasZbv:bnfp4nl)*
<!-- Sent from Reviewable.io -->

benesch commented:
PTAL. Now, rather than removing `all`, this makes `all` an alias for `build`. This is consistent with the GNU standard Make targets. (`build` is non-standard.)

dt reviewed (approved):
🐶

a-robinson reviewed (approved):
I like this change 👍

BramGruneir commented:
Ok, so I like just running make and knowing that it covers most things.  What's my new workflow now?  Do I need to be explicit?

bdarnell commented:
Yeah, there should be just one target to include all the stuff you should run before committing. I think `make check` should become an alias for `make build test lint` (instead of just `make test` as proposed in #15866). 

And while we're renaming things, the standard practice would be to name the `build` target after its output, i.e. `cockroach`. But maybe that would be misleading since there's no proper dependency tracking and we'd have to make that target `.PHONY`.

benesch commented:
Agreed, `build` should would ideally be named after the binary it produces. That binary is named `cockroach$(SUFFIX)$$($(XGO) env GOEXE)`, though, so it's somewhat of a moving target. I think we'll always want to provide `build` as a convenience.

benesch committed:
build: add pre-push target that runs all linters and tests

Importantly, this makes it possible to run the UI linters and unit tests
along with the Go linters and unit tests in one command.

benesch commented:
Ok, PTAL. Here's a summary:

  * `make` →  `make all`
  * `make all` → `make build`
  * `make pre-push` → `make generate lint test && cd pkg/ui make lint test && `
  * `make check` → "did you mean `make lint` or `make test`?"

I'm holding off on changing the behavior of `make check` for now, since there's still no consensus on whether it should be an alias for `test` or `pre-push`. In the meantime, it seems everyone agrees `make` should mean `make build`, provided there's a shortcut to run everything. `make pre-push` is exactly that shortcut. Feel free to bikeshed the name, though I think `pre-push` is descriptive, short, and unburdened with Make convention. 

@BramGruneir does 5m for `pre-push` sound about right?

```
$ time make pre-push
...
real	4m53.608s
user	16m49.628s
sys	1m58.850s
```

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15909#-:-KkRRsu4ZZQ01ZFUfuNZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Yeah, sounds good to me.  5 mins locally makes sense and give me an excuse to get a faster machine.

bdarnell commented:
LGTM. `make testrace` is a part of my pre-push routine, although it adds a lot to the runtime so maybe it's not worth adding to `make pre-push`. 

I retract my earlier recommendation that `make check` be equivalent to `pre-push` (was I the only one in this camp?). I now vote for it to be an alias for `make test` (or maybe `make test && make -C pkg/ui test`).

benesch commented:
🎉 TFTRs, everyone! In this goes.

I know of at least a few folks who can't run `testrace` locally without pain, so I'm going to leave that out for now.
